File: source/extensions/anymal.ros1_cpp.pub_node/plugins/anymal.ros1_cpp.pub_node/ANYmalPybindExtension.cpp
// Copyright (c) 2022, NVIDIA CORPORATION. All rights reserved.
//
// NVIDIA CORPORATION and its licensors retain all intellectual property
// and proprietary rights in and to this software, related documentation
// and any modifications thereto.  Any use, reproduction, disclosure or
// distribution of this software and related documentation without an express
// license agreement from NVIDIA CORPORATION is strictly prohibited.
//

#define CARB_EXPORTS

#include <carb/PluginUtils.h>

#include <omni/ext/IExt.h>

#include <anymal/ros1_cpp/pub_node/ANYmalBoundInterface.h>

#include <unordered_map>

const struct carb::PluginImplDesc pluginImplDesc = { "anymal.ros1_cpp.pub_node.plugin",
                                                     "C++ extension to publish ros topic for ANYmal.", "RSL",
                                                     carb::PluginHotReload::eEnabled, "dev" };

namespace anymal
{
namespace ros1_cpp
{
namespace pub_node
{

class ANYmalBoundImplementation : public ANYmalBoundInterface
{
public:
    void registerBoundObject(carb::ObjectPtr<ANYmalBoundObject>& object) override
    {
        if (object)
        {
            m_registeredObjectsById[object->getId()] = object;
        }
    }

    void deregisterBoundObject(carb::ObjectPtr<ANYmalBoundObject>& object) override
    {
        if (object)
        {
            const auto& it = m_registeredObjectsById.find(object->getId());
            if (it != m_registeredObjectsById.end())
            {
                m_registeredObjectsById.erase(it);
            }
        }
    }

    carb::ObjectPtr<ANYmalBoundObject> findBoundObject(const char* id) const override
    {
        const auto& it = m_registeredObjectsById.find(id);
        if (it != m_registeredObjectsById.end())
        {
            return it->second;
        }

        return carb::ObjectPtr<ANYmalBoundObject>();
    }

private:
    std::unordered_map<std::string, carb::ObjectPtr<ANYmalBoundObject>> m_registeredObjectsById;
};

}
}
}

CARB_PLUGIN_IMPL(pluginImplDesc, anymal::ros1_cpp::pub_node::ANYmalBoundImplementation)

void fillInterface(anymal::ros1_cpp::pub_node::ANYmalBoundImplementation& iface)
{
}

File: source/extensions/anymal.ros1_cpp.pub_node/python/impl/anymal_pybind_extension.py
## Copyright (c) 2022, NVIDIA CORPORATION.  All rights reserved.
##
## NVIDIA CORPORATION and its licensors retain all intellectual property
## and proprietary rights in and to this software, related documentation
## and any modifications thereto.  Any use, reproduction, disclosure or
## distribution of this software and related documentation without an express
## license agreement from NVIDIA CORPORATION is strictly prohibited.
##
import omni.ext
from .._anymal_pybind_bindings import *

# Global public interface object.
_bound_interface = None

# Public API.
def get_bound_interface() -> ANYmalBoundInterface:
    return _bound_interface


# Use the extension entry points to acquire and release the interface.
class ANYmalPybindExtension(omni.ext.IExt):
    def __init__(self):
        super().__init__()

        global _bound_interface
        _bound_interface = acquire_bound_interface()

    def on_shutdown(self):
        global _bound_interface
        release_bound_interface(_bound_interface)
        _bound_interface = None
